skimage/util/compare.py

```diff
@@ -1,27 +1,143 @@
+        if parameters['image2'].default is not DEPRECATED:
+            raise RuntimeError(
+                f"Expected `{self.deprecated_name}` to have the value {DEPRECATED!r} "
+                f"to indicate its status in the rendered signature."
+            )
+        warning_message = (
+            "Since version 0.23, the two input images are named `image0` and "
+            "`image1` (instead of `image1` and `image2`, respectively). Please use "
+            "`image0, image1` to avoid this warning for now, and avoid an error "
+            "from version 0.25 onwards."
+        )
+        wm_method = (
+            "Starting in version 0.25, all arguments following `image0, image1` "
+            "(including `method`) will be keyword-only. Please pass `method=` "
+            "in the function call to avoid this warning for now, and avoid an error "
+            "from version 0.25 onwards."
+        )
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if 'image2' not in kwargs.keys():
+                kwargs['image2'] = DEPRECATED
+
+            # Pass first all args as kwargs
+            if len(args) > 0:
+                kwargs['image0'] = args[0]
+                if len(args) > 1:
+                    kwargs['image1'] = args[1]
+                    if len(args) > 2 and args[len(args) - 1] in [
+                        'diff',
+                        'blend',
+                        'checkerboard',
+                    ]:
+                        warnings.warn(wm_method, category=FutureWarning)
+                        kwargs['method'] = args[len(args) - 1]
+
+            if kwargs['image2'] is not DEPRECATED:
+                deprecated_value = kwargs['image2']
+                kwargs['image2'] = DEPRECATED
+                if 'image1' in kwargs.keys():
+                    if 'image0' in kwargs.keys():
+                        raise RuntimeError(
+                            "Use `image0, image1` to pass the two input images."
+                        )
+                    else:
+                        warnings.warn(warning_message, category=FutureWarning)
+                        args = (kwargs['image1'], deprecated_value)
+                else:
+                    if 'image0' in kwargs.keys():
+                        warnings.warn(warning_message, category=FutureWarning)
+                        args = (kwargs['image0'], deprecated_value)
+
+            kwargs.pop('image2')
+            if 'image0' in kwargs.keys():
+                kwargs.pop('image0')
+            if 'image1' in kwargs.keys():
+                kwargs.pop('image1')
+            return func(*args, **kwargs)
+
+        return wrapper
 
 
-def compare_images(image1, image2, method='diff', *, n_tiles=(8, 8)):
+@_rename_image_params("image2", start_version="0.23", stop_version="0.25")
+def compare_images(image0, image1, image2=DEPRECATED, method='diff', *, n_tiles=(8, 8)):
```

Comment:
I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things. And also breaks

```python
a = np.ones(10)
b = np.zeros(10)
ski.util.compare_images(a, b, "blend")
# warns but still uses method="diff"
```

which used to work before and should continue working during the deprecation. I think we can't use an argument with `=DEPRECATED` here.

Except for that I think you got quite close, especially with the idea to first turn all `args` into `kwargs` to more easily handle them. What do you think about the draft and decorator below? It should work for all cases unless I've forgotten a a way to abuse the signature. It also turns `method` into a keyword-only parameter.

```python
def _rename_image_params(func):

    @functools.wraps(func)
    def wrapper(*args, **kwargs):

        # Turn all into keyword parameters
        for i, (value, param) in enumerate(
            zip(args, ["image0", "image1", "method", "n_tiles"])
        ):
            if i >= 2:
                warnings.warn("method will become keyword-only")
            if param in kwargs:
                raise ValueError(
                    f"{param} given as positional and keyword argument"
                )
            else:
                kwargs[param] = value
        args = tuple()

        # Account for `image2` if given
        if "image2" in kwargs:
            warnings.warn("used deprecated image2")

            # Safely move `image2` to `image1` if that's empty
            if "image1" in kwargs:
                # Safely move `image1` to `image0`
                if "image0" in kwargs:
                    raise ValueError(
                        "three images given, `image0`, `image1` and `image2`"
                    )
                kwargs["image0"] = kwargs.pop("image1")
            kwargs["image1"] = kwargs.pop("image2")

        return func(*args, **kwargs)

    return wrapper
```

Of course it doesn't have the nice convenience stuff of modifying the docstring and warnings and error messages are drafts and need to be polished. But I think this deprecation is so special that we should just edit the docstring by hand.

Reply:
> I'm sorry but unfortunately I think looking at `deprecate_parameter` has lead you on a wrong path and over-complicated things.

No worries! I did this quickly and thought I would iterate afterwards.

> And also breaks
> 
> ```python
> a = np.ones(10)
> b = np.zeros(10)
> ski.util.compare_images(a, b, "blend")
> # warns but still uses method="diff"
> ```
> 
> which used to work before and should continue working during the deprecation.

Pardon me? I just tried locally and it works; it warns, indeed, and it returns
```py
array([0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5])
```
just like current main. `method="diff"` returns
 ```py
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
```
instead. I just noticed that `method="checkerboard"` errors because `a` and `b` are 1-dimensional, which doesn't work with
https://github.com/scikit-image/scikit-image/blob/e453a53d23a980408365d35579008a588868ffe0/skimage/util/compare.py#L48

So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 :grimacing:

Reply:
> So we should probably improve the function by checking, in the `method="checkerboard"` case, that ndim == 2 😬

Done db1b7740e21637a3. If you prefer, I can cherry-pick this commit and submit it as a separate PR.

Reply:
Ah, you are right, it does handle this. I chose my example poorly and got confused by the `image2=DEPRECATED` as the third positional arg. 

... which brings me to the point that we should probably remove it and make it so that the function shows the expected new signature as un-ambiguously as possible.